File: core/cetech/component.lua
local ffi = require("ffi")
local api_system  = require("cetech/api_system")
local hash_lib  = require("cetech/hashlib")

ffi.cdef[[
struct ct_blob;

typedef struct {
    void* v;
} yaml_node_t;

typedef int (*ct_component_compiler_t)(yaml_node_t body,
                                       struct ct_blob *data);


enum ct_property_type {
    PROPERTY_INVALID = 0,
    PROPERTY_BOOL,
    PROPERTY_FLOAT,
    PROPERTY_STRING,
    PROPERTY_STRINGID64,
    PROPERTY_VEC3,
    PROPERTY_QUATF,
};

struct ct_property_value {
    enum ct_property_type type;

    union {
        int b;
        float f;
        const char *str;
        float vec3f[3];
        float quatf[4];
        uint64_t strid_64;
    } value;
};

static struct ct_component_clb {
    void (*destroyer)(struct ct_world world,
                      struct ct_entity *ents,
                      size_t ent_count);

    void (*spawner)(struct ct_world world,
                    struct ct_entity *ents,
                    uint32_t *cent,
                    uint32_t *ents_parent,
                    size_t ent_count,
                    void *data);

    void (*set_property)(struct ct_world world,
                         struct ct_entity entity,
                         uint64_t key,
                         struct ct_property_value value);

    struct ct_property_value (*get_property)(struct ct_world world,
                                             struct ct_entity entity,
                                             uint64_t key);
    ct_world_callbacks_t world_clb;
};


struct ct_component_a0 {
    void (*register_compiler)(uint64_t type,
                              ct_component_compiler_t compiler,
                              uint32_t spawn_order);

    int (*compile)(uint64_t type,
                   yaml_node_t body,
                   struct ct_blob *data);

    uint32_t (*spawn_order)(uint64_t type);

    void (*register_type)(uint64_t type,
                          struct ct_component_clb clb);

    void (*spawn)(struct ct_world world,
                  uint64_t type,
                  struct ct_entity *ent_ids,
                  uint32_t *cent,
                  uint32_t *ents_parent,
                  uint32_t ent_count,
                  void *data);

    void (*destroy)(struct ct_world world,
                    struct ct_entity *ent,
                    uint32_t count);

    void (*set_property)(uint64_t type,
                         struct ct_world world,
                         struct ct_entity entity,
                         uint64_t key,
                         struct ct_property_value value);

    struct ct_property_value (*get_property)(uint64_t type,
                                             struct ct_world world,
                                             struct ct_entity entity,
                                             uint64_t key);
};
]]

local C = ffi.C

local api = api_system.load("ct_component_a0")

Component = {}

function Component.set_property(world, entity, component_type, key, value)
    local component_type = hash_lib.id64_from_str(component_type)
    local key_id = hash_lib.id64_from_str(key)

    local value_type = type(value)
    local prop_value

    if value_type == 'number' then
        prop_value = ffi.new("struct ct_property_value", { C.PROPERTY_FLOAT, { f=value }})
    elseif value_type == 'string' then
        prop_value = ffi.new("struct ct_property_value", { C.PROPERTY_STRING, { str=value }})
    elseif value_type == 'boolean' then
        prop_value = ffi.new("struct ct_property_value", { C.PROPERTY_BOOL, { b=value }})
    else
        if cetech.Vec3f.is(value) then
            prop_value = ffi.new("struct ct_property_value", { C.PROPERTY_VEC3, { vec3f=ffi.cast("float*", value)[0] }})
        elseif cetech.Quatf.is(value) then
            prop_value = ffi.new("struct ct_property_value", { C.PROPERTY_QUATF, { quatf=ffi.cast("float*", value)[0] }})
        end
    end

    return api.set_property(component_type, world, entity, key_id, prop_value)
end

function Component.get_property(world, entity, component_type, key)
    local component_type = hash_lib.id64_from_str(component_type)
    local key_id = hash_lib.id64_from_str(key)

    local prop_value = api.get_property(component_type, world, entity, key_id)

    if prop_value.type == C.PROPERTY_FLOAT then
        return prop_value.value.f

    elseif prop_value.type == C.PROPERTY_STRING then
        return prop_value.value.str

    elseif prop_value.type == C.PROPERTY_BOOL then
        return prop_value.value.b

    elseif prop_value.type == C.PROPERTY_VEC3 then
        local v = prop_value.value.vec3f
        return cetech.Vec3f.make(v[0], v[1], v[2])

    elseif prop_value.type == C.PROPERTY_QUATF then
        local q = prop_value.value.quatf
        return cetech.Quatf.make(q[0], q[1], q[2], q[3])
    end
end

return Component

File: src/cetech/entity/private/component.cpp
//==============================================================================
// Includes
//==============================================================================

#include <cetech/entity/entity.h>
#include <celib/hash.h>
#include "celib/map.inl"

#include "cetech/os/memory.h"
#include "cetech/config/config.h"
#include "cetech/api/api_system.h"
#include "cetech/module/module.h"

CETECH_DECL_API(ct_memory_a0);
CETECH_DECL_API(ct_world_a0);

using namespace celib;

//==============================================================================
// Globals
//==============================================================================

#define _G ComponentMaagerGlobals
static struct _G {
    cel_hash_t spawn_order_map;

    Map<ct_component_compiler_t> compiler_map;
    Map<ct_component_clb> component_clb;

    cel_alloc* allocator;
} _G;


//==============================================================================
// Public interface
//==============================================================================

static void register_compiler(uint64_t type,
                       ct_component_compiler_t compiler,
                       uint32_t spawn_order) {
    map::set(_G.compiler_map, type, compiler);

    cel_hash_add(&_G.spawn_order_map, type, spawn_order, _G.allocator);
}

static int compile(uint64_t type,
            const char *filename,
            uint64_t *component_key,
            uint32_t component_key_count,
            char **data) {

    ct_component_compiler_t compiler = map::get<ct_component_compiler_t>(
            _G.compiler_map, type, nullptr);

    if (!compiler) {
        return 0;
    }

    return compiler(filename, component_key, component_key_count, data);
}

static uint32_t get_spawn_order(uint64_t type) {
    return (uint32_t) cel_hash_lookup(&_G.spawn_order_map, type,  0);
}

static void register_type(uint64_t type,
                   ct_component_clb clb) {
    map::set(_G.component_clb, type, clb);

    ct_world_callbacks_t wclb = {
            .on_created = clb.world_clb.on_created,
            .on_destroy = clb.world_clb.on_destroy,
            .on_update = clb.world_clb.on_update,
    };

    ct_world_a0.register_callback(wclb);
}

static void spawn(ct_world world,
           uint64_t type,
           ct_entity *ent_ids,
           uint32_t *cent,
           uint32_t *ents_parent,
           uint32_t ent_count,
           void *data) {

    ct_component_clb clb = map::get(_G.component_clb, type,
                                    ct_component_clb_null);

    if (!clb.spawner) {
        return;
    }

    clb.spawner(world, ent_ids, cent, ents_parent, ent_count, data);
}

static void destroy(ct_world world,
             ct_entity *ent,
             uint32_t count) {

    auto ct_it = map::begin(_G.component_clb);
    auto ct_end = map::end(_G.component_clb);

    while (ct_it != ct_end) {
        ct_it->value.destroyer(world, ent, count);
        ++ct_it;
    }
}

static ct_component_a0 component_api = {
        .register_compiler = register_compiler,
        .compile = compile,
        .spawn_order = get_spawn_order,
        .register_type = register_type,
        .spawn = spawn,
        .destroy = destroy,
};

static void _init_api(ct_api_a0 *a0) {
    a0->register_api("ct_component_a0", &component_api);
}

static void _init(ct_api_a0 *a0) {
    _init_api(a0);

    _G = {
            .allocator = ct_memory_a0.main_allocator()
    };

    _G.compiler_map.init(ct_memory_a0.main_allocator());
    _G.component_clb.init(ct_memory_a0.main_allocator());
}

static void _shutdown() {
    _G.compiler_map.destroy();
    _G.component_clb.destroy();
    cel_hash_free(&_G.spawn_order_map, _G.allocator);
}

CETECH_MODULE_DEF(
        component,
        {
            CETECH_GET_API(api, ct_memory_a0);
            CETECH_GET_API(api, ct_world_a0);
        },
        {
            CEL_UNUSED(reload);
            _init(api);
        },
        {
            CEL_UNUSED(reload);
            CEL_UNUSED(api);
            _shutdown();
        }
)

